contentcuration/contentcuration/utils/recommendations.py

```diff
@@ -61,67 +76,307 @@ def create_backend(self) -> Backend:
+                    rank=node['rank'],
+                    override_threshold=override_threshold,
+                ) for node in nodes if node['contentnode_id'] not in existing_cache
+            ]
+            RecommendationsCache.objects.bulk_create(new_cache, ignore_conflicts=True)
+            return True
+        except Exception as e:
+            logging.exception(e)
+            return False
+
+    def _extract_override_threshold(self, request) -> bool:
+        """
+        Extracts the override_threshold parameter from the request safely.
+
+        :param request: The request containing the parameters.
+        :return: The value of the override_threshold parameter, or False if not present.
+        :rtype: bool
+        """
+        return request.params.get('override_threshold', False) if request.params else False
+
+    def get_recommendations(self, topic: Dict[str, Any],
+                            override_threshold=False) -> RecommendationsResponse:
+        """
+        Get recommendations for the given topic.
+
+        :param topic: A dictionary containing the topic for which to get recommendations. See
+        https://github.com/learningequality/le-utils/blob/main/spec/schema-embed_topics_request.json
+        :param override_threshold: A boolean flag to override the recommendation threshold.
+        :return: The recommendations for the given topic. :rtype: RecommendationsResponse
+        """
+
+        recommendations = []
+        request = EmbedTopicsRequest(
+            params={'override_threshold': override_threshold},
+            json=topic,
+        )
+
+        cached_response = self.response_exists(request)
+        if cached_response:
+            response = cached_response
+        else:
+            response = self.generate_embeddings(request=request)
+            if not response.error:
+                self.cache_embeddings_request(request, response)
+
+        nodes = self._extract_data(response)
+        if len(nodes) > 0:
+            node_ids = [node['contentnode_id'] for node in nodes]
+
+            # Get the channel_id from PublicContentNode based on matching node_id from ContentNode
+            channel_id_subquery = PublicContentNode.objects.filter(
+                self._normalize_uuid(F('id')) == self._normalize_uuid(OuterRef('node_id'))
+            ).values('channel_id')[:1]
+
+            # Get main_tree_id from Channel based on channel_id obtained from channel_id_subquery
+            main_tree_id_subquery = Channel.objects.filter(
+                self._normalize_uuid(F('id')) == self._normalize_uuid(Subquery(channel_id_subquery))
+            ).values('main_tree_id')[:1]
+
+            # Annotate main_tree_id onto ContentNode
+            recommendations = ContentNode.objects.filter(id__in=node_ids).annotate(
+                main_tree_id=Subquery(main_tree_id_subquery)
+            ).values('id', 'node_id', 'main_tree_id', 'parent_id')
```

Comment:
This does not look correct to me. Are there tests covering this code?

`channel_id_subquery` has an `OuterRef` to `node_id` but is used in a `Channel` queryset subquery, which means the `OuterRef` should apply to the `Channel` model, which doesn't have a `node_id`.

Reiterating again that the `PublicContentNode.id` is the same as `ContentNode.node_id`.